File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-01-decision-framework/decision-framework-tool-java/src/main/java/com/amplitudeseguros/decisionframework/Ahp.java
package com.amplitudeseguros.decisionframework;

import java.util.List;

/**
 * AHP -- Analytic Hierarchy Process (Saaty 1980). Deriva pesos de uma matriz de
 * comparação pareada pelo método da média geométrica das linhas, e calcula a
 * Razão de Consistência (CR) do julgamento. Equivalente a derivarPesosAHP,
 * calcularConsistenciaAHP e agregarMatrizesComite em decision-framework-tool.js.
 */
public final class Ahp {

    /** Random Index de Saaty (n=4) -- normaliza o Índice de Consistência em Razão de Consistência. */
    public static final double RANDOM_INDEX_N4 = 0.90;

    private Ahp() {
    }

    public record Consistencia(double lambdaMax, double ci, double cr, boolean consistente) {
    }

    /** Vetor de prioridades (pesos), soma 1.0, pela média geométrica das linhas (row geometric mean method). */
    public static double[] derivarPesos(double[][] matriz) {
        int n = matriz.length;
        double[] mediasGeometricas = new double[n];
        for (int i = 0; i < n; i++) {
            double produto = 1;
            for (double v : matriz[i]) {
                produto *= v;
            }
            mediasGeometricas[i] = Math.pow(produto, 1.0 / n);
        }
        double soma = 0;
        for (double v : mediasGeometricas) {
            soma += v;
        }
        double[] pesos = new double[n];
        for (int i = 0; i < n; i++) {
            pesos[i] = mediasGeometricas[i] / soma;
        }
        return pesos;
    }

    /** CR < 0.10 (limiar padrão de Saaty) considera o julgamento consistente o bastante para confiar nos pesos. */
    public static Consistencia calcularConsistencia(double[][] matriz, double[] pesos) {
        int n = matriz.length;
        double[] aw = new double[n];
        for (int i = 0; i < n; i++) {
            double soma = 0;
            for (int j = 0; j < n; j++) {
                soma += matriz[i][j] * pesos[j];
            }
            aw[i] = soma;
        }
        double somaRazoes = 0;
        for (int i = 0; i < n; i++) {
            somaRazoes += aw[i] / pesos[i];
        }
        double lambdaMax = somaRazoes / n;
        double ci = (lambdaMax - n) / (n - 1);
        double cr = ci / RANDOM_INDEX_N4;
        return new Consistencia(lambdaMax, ci, cr, cr < 0.10);
    }

    /**
     * Agrega várias matrizes de comparação pareada (uma por avaliador) pela média
     * geométrica célula a célula (método AIJ) -- preserva a propriedade recíproca
     * (a[j][i] = 1/a[i][j]), então o resultado ainda é uma matriz de comparação
     * pareada válida. Um comitê de 1 avaliador reduz ao caso original.
     */
    public static double[][] agregarMatrizesComite(List<double[][]> matrizes) {
        int nAvaliadores = matrizes.size();
        int n = matrizes.get(0).length;
        double[][] agregada = new double[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                double produto = 1;
                for (double[][] matriz : matrizes) {
                    produto *= matriz[i][j];
                }
                agregada[i][j] = Math.pow(produto, 1.0 / nAvaliadores);
            }
        }
        return agregada;
    }
}

File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-01-decision-framework/decision-framework-tool-go/ahp/ahp.go
// Package ahp implementa o Analytic Hierarchy Process (Saaty, 1980): deriva
// pesos de uma matriz de comparação pareada pela média geométrica das linhas,
// calcula a Razão de Consistência (CR) do julgamento, e agrega o julgamento de
// vários avaliadores (comitê) numa única matriz. Equivalente a derivarPesosAHP,
// calcularConsistenciaAHP e agregarMatrizesComite em decision-framework-tool.js.
package ahp

import "math"

// RandomIndexN4 é o Random Index de Saaty (n=4), usado pra normalizar o
// Índice de Consistência em Razão de Consistência.
const RandomIndexN4 = 0.90

// Consistencia é o resultado de calcularConsistencia.
type Consistencia struct {
	LambdaMax   float64
	CI          float64
	CR          float64
	Consistente bool
}

// DerivarPesos retorna o vetor de prioridades (pesos, soma 1.0) pelo método da
// média geométrica das linhas (row geometric mean method).
func DerivarPesos(matriz [][]float64) []float64 {
	n := len(matriz)
	mediasGeometricas := make([]float64, n)
	for i, linha := range matriz {
		produto := 1.0
		for _, v := range linha {
			produto *= v
		}
		mediasGeometricas[i] = math.Pow(produto, 1.0/float64(n))
	}
	soma := 0.0
	for _, v := range mediasGeometricas {
		soma += v
	}
	pesos := make([]float64, n)
	for i, v := range mediasGeometricas {
		pesos[i] = v / soma
	}
	return pesos
}

// CalcularConsistencia calcula a Razão de Consistência (CR); CR < 0.10 é o
// limiar padrão de Saaty pra considerar o julgamento consistente.
func CalcularConsistencia(matriz [][]float64, pesos []float64) Consistencia {
	n := len(matriz)
	aw := make([]float64, n)
	for i, linha := range matriz {
		soma := 0.0
		for j, v := range linha {
			soma += v * pesos[j]
		}
		aw[i] = soma
	}
	somaRazoes := 0.0
	for i := range aw {
		somaRazoes += aw[i] / pesos[i]
	}
	lambdaMax := somaRazoes / float64(n)
	ci := (lambdaMax - float64(n)) / float64(n-1)
	cr := ci / RandomIndexN4
	return Consistencia{LambdaMax: lambdaMax, CI: ci, CR: cr, Consistente: cr < 0.10}
}

// AgregarMatrizesComite agrega várias matrizes de comparação pareada (uma por
// avaliador) pela média geométrica célula a célula (método AIJ) -- preserva a
// propriedade recíproca (a[j][i] = 1/a[i][j]), então o resultado ainda é uma
// matriz de comparação pareada válida. Um comitê de 1 avaliador reduz ao caso
// original.
func AgregarMatrizesComite(matrizes [][][]float64) [][]float64 {
	nAvaliadores := len(matrizes)
	n := len(matrizes[0])
	agregada := make([][]float64, n)
	for i := 0; i < n; i++ {
		agregada[i] = make([]float64, n)
		for j := 0; j < n; j++ {
			produto := 1.0
			for _, matriz := range matrizes {
				produto *= matriz[i][j]
			}
			agregada[i][j] = math.Pow(produto, 1.0/float64(nAvaliadores))
		}
	}
	return agregada
}
